controller/controller.go

```diff
@@ -1026,6 +1063,14 @@ func (ctrl *ProvisionController) provisionClaimOperation(claim *v1.PersistentVol
 		}
 		err = fmt.Errorf("failed to provision volume with StorageClass %q: %v", claimClass, err)
 		ctrl.eventRecorder.Event(claim, v1.EventTypeWarning, "ProvisioningFailed", err.Error())
+		if _, ok := claim.Annotations[annSelectedNode]; ok {
+			// For dynamic PV provisioning with delay binding, provisioner may fail
+			// because node is wrong. `selectedNode` must be removed to notify
+			// scheduler to schedule again.
+			// No need to handle error here, because if `selectedNode` is failed to
+			// be removed, next retry will remove it again.
+			ctrl.rescheduleProvisioning(claim)
```

Comment:
We only get here if the provisioner really returns a permanent provisioning failure, right?

Right now, external-provisioner will not do that if the CSI driver returns a non-final error:
https://github.com/kubernetes-csi/external-provisioner/blob/6019c43382549945cf7b80f1bffa826c6d14392c/pkg/controller/controller.go#L1161-L1183

If a CSI driver returns `codes.ResourceExhausted`, we want to stop provisioning on the node and try elsewhere. For that to work, we probably need an additional parameter that tells the provisioner that it is expected to give up in that case.

Reply:
Nope, this is the code that handles any kind of error, including timeouts where the provisioner must continue to attempt provisioning to avoid leaking volumes.

Apparently a new API with richer semantic was introduced a while back that the lib should be using instead:
https://github.com/kubernetes-csi/external-provisioner/blob/6019c43382549945cf7b80f1bffa826c6d14392c/pkg/controller/controller.go#L395-L399